src/github.com/matrix-org/dendrite/clientapi/readers/register.go

```diff
@@ -0,0 +1,66 @@
+package readers
+
+import (
+	"fmt"
+	"github.com/matrix-org/dendrite/clientapi/auth/storage/accounts"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/util"
+	"net/http"
+	"regexp"
+)
+
+const (
+	maxUsernameLength = 254 // http://matrix.org/speculator/spec/HEAD/intro.html#user-identifiers TODO account for domain
+)
+
+var validUsernameRegex = regexp.MustCompile(`^[0-9a-zA-Z_\-./]+$`)
+
+func validate(username string) *util.JSONResponse {
+	if len(username) > maxUsernameLength {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.BadJSON(fmt.Sprintf("'username' >%d characters", maxUsernameLength)),
+		}
+	} else if !validUsernameRegex.MatchString(username) {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.InvalidUsername("User ID can only contain characters a-z, 0-9, or '_-./'"),
+		}
+	} else if username[0] == '_' { // Regex checks its not a zero length string
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.InvalidUsername("User ID can't start with a '_'"),
+		}
+	}
+	return nil
+}
```

Comment:
This seems to be duplicated from writers.register? Either we should make he validation function in writers public, or we should move the RegisterAvailable func to writers.register (which is probably my preferred approach)

Reply:
The function on writers.register is basicly te same except that one also checks the password. So either way I would atleast need to split the validate function in username and password

Reply:
I think splitting up into `validateUserName` and `validatePassword` makes sense